sdk/signalr/azure-messaging-webpubsubservice/azure/messaging/webpubsubservice/__init__.py

```diff
@@ -0,0 +1,188 @@
+        """Create a new WebPubSubServiceClient instance
+
+        :param endpoint: Endpoint to connect to.
+        :type endpoint: ~str
+        :param credential: Credentials to use to connect to endpoint.
+        :type credential: ~azure.core.credentials.AzureKeyCredentials
+        :keyword api_version: Api version to use when communicating with the service.
+        :type api_version: str
+        :keyword user: User to connect as. Optional.
+        :type user: ~str
+        """
+        self.endpoint = endpoint.rstrip("/")
+        transport = kwargs.pop("transport", None) or coretransport.RequestsTransport(
+            **kwargs
+        )
+        kwargs.setdefault('sdk_moniker', 'messaging-webpubsubservice/{}'.format(_VERSION))
+        policies = [
+            corepolicies.HeadersPolicy(**kwargs),
+            corepolicies.UserAgentPolicy(**kwargs),
+            corepolicies.RetryPolicy(**kwargs),
+            corepolicies.ProxyPolicy(**kwargs),
+            corepolicies.CustomHookPolicy(**kwargs),
+            corepolicies.RedirectPolicy(**kwargs),
+            JwtCredentialPolicy(credential, kwargs.get("user", None)),
+            corepolicies.NetworkTraceLoggingPolicy(**kwargs),
+        ]  # type: Any
+        self._pipeline = corepipeline.Pipeline(
+            transport,
+            policies,
+        )  # type: corepipeline.Pipeline
+
+
+    @classmethod
+    def from_connection_string(cls, connection_string, **kwargs):
+        # type: (Type[ClientType], str, Any) -> ClientType
+        for invalid_keyword_arg in ('endpoint', 'accesskey'):
+            if invalid_keyword_arg in kwargs:
+                raise TypeError('Unknown argument {}'.format(invalid_keyword_arg))
+
+        for segment in connection_string.split(";"):
+            if '=' in segment:
+                key, value = segment.split('=', maxsplit=1)
+                key = key.lower()
+                if key == 'version':
+                    # The name in the connection string != the name for the constructor.
+                    # Let's map it to whatthe constructor actually wants...
+                    key = 'api_version'
+                kwargs[key] = value
+            else:
+                raise ValueError("Malformed connection string - expected 'key=value', got {}".format(segment))
+
+        if 'endpoint' not in kwargs:
+            raise ValueError("connection_string missing 'endpoint' field")
+
+        if 'accesskey' not in kwargs:
+            raise ValueError("connection_string missing 'accesskey' field")
+
+        kwargs['credential'] = corecredentials.AzureKeyCredential(kwargs.pop('accesskey'))
+        return cls(**kwargs)
+
+    def __repr__(self):
+        return "<WebPubSubServiceClient> endpoint:'{}'".format(self.endpoint)
+
+    def _format_url(self, url):
+        # type: (str) -> str
+        assert self.endpoint[-1] != "/", "My endpoint should not have a trailing slash"
```

Comment:
Would be kind of strange to encounter an `AssertError` when constructing the client?

Reply:
In that scenario, this is truly an assert (we are stripping the trailing slash before calling the method). However, when looking at the code, the endpoint is a read/write attribute (which it probably shouldn't be) and the user could have set it to a value that does not conform to our assumptions. I may want to change self.endpoint to a read-only property...